src/random_ray/iteration.cpp

```diff
@@ -0,0 +1,610 @@
+  double volume_normalization_factor =
+    1.0 / (total_active_distance_per_iteration * simulation::current_batch);
+
+// Normalize Scalar flux to total distance travelled by all rays this iteration
+#pragma omp parallel for
+  for (int64_t e = 0; e < random_ray::scalar_flux_new.size(); e++) {
+    random_ray::scalar_flux_new[e] *= normalization_factor;
+  }
+
+// Accumulate cell-wise ray length tallies collected this iteration, then
+// update the simulation-averaged cell-wise volume estimates
+#pragma omp parallel for
+  for (int64_t sr = 0; sr < random_ray::n_source_regions; sr++) {
+    random_ray::volume_t[sr] += random_ray::volume[sr];
+    random_ray::volume[sr] =
+      random_ray::volume_t[sr] * volume_normalization_factor;
+  }
+}
+
+// Combines transport flux contributions and flat source contributions
+// from the previous iteration to generate this iteration's estimate of
+// scalar flux.
+int64_t add_source_to_scalar_flux()
+{
+  int negroups = data::mg.num_energy_groups_;
+
+  int64_t n_hits = 0;
+
+  // Temperature and angle indices, if using multiple temperature
+  // data sets and/or anisotropic data sets.
+  // TODO: Currently assumes we are only using single temp/single
+  // angle data.
+  const int t = 0;
+  const int a = 0;
+
+#pragma omp parallel for reduction(+ : n_hits)
+  for (int sr = 0; sr < random_ray::n_source_regions; sr++) {
+
+    // Check if this cell was hit this iteration
+    int was_cell_hit = random_ray::was_hit[sr];
+    if (was_cell_hit) {
+      n_hits++;
+    }
+
+    double volume = random_ray::volume[sr];
+    int material = random_ray::material[sr];
+    for (int e = 0; e < negroups; e++) {
+      int64_t idx = (sr * negroups) + e;
+
+      // There are three scenarios we need to consider:
+      if (was_cell_hit) {
+        // 1. If the FSR was hit this iteration, then the new flux is equal to
+        // the flat source from the previous iteration plus the contributions
+        // from rays passing through the source region (computed during the
+        // transport sweep)
+        float Sigma_t = data::mg.macro_xs_[material].get_xs(
+          MgxsType::TOTAL, e, nullptr, nullptr, nullptr, t, a);
+        random_ray::scalar_flux_new[idx] /= (Sigma_t * volume);
+        random_ray::scalar_flux_new[idx] += random_ray::source[idx];
+      } else if (volume > 0.0) {
+        // 2. If the FSR was not hit this iteration, but has been hit some
+        // previous iteration, then we simply set the new scalar flux to be
+        // equal to the contribution from the flat source alone.
+        random_ray::scalar_flux_new[idx] = random_ray::source[idx];
+      } else {
+        // If the FSR was not hit this iteration, and it has never been hit in
+        // any iteration (i.e., volume is zero), then we want to set this to 0
+        // to avoid dividing anything by a zero volume.
+        random_ray::scalar_flux_new[idx] = 0.f;
+      }
+    }
```

Comment:
This is a very interesting approach, and not something we looked at in Ben's transport theory class. I always assumed you need to run enough rays to hit each source region, so this is very neat!

Reply:
Yes, it's surprising how many FSRs you can miss and get away with! It remains stable even missing like 30% of the FSRs each iteration (though becomes a bit inefficient at that point due to static iteration costs).

Reply:
This should be mentioned in some documentation, since IIRC this was not specifically highlighted in your JCP paper! It's very important for a user to consider when setting up the calculation.

Reply:
This is now discussed in the docs (in the user guide -> random ray section -> ray density subsection).